Migrate KafkaJS to ^2.0.0

kalyvasio opened the pull request:
Bumps KafkaJS to 2.0.0

It contains two major changes that affect us

## Manage topics through admin client
KafkaJS contains some non-backwards compatible options in regards to [managing topics via the admin client](https://kafka.js.org/docs/migration-guide-v2.0.0#admin-fetchoffsets-accepts-topics-instead-of-topic)
We have changed the [implementation](https://github.com/Workable/orka/pull/384/files#diff-7a49104279a769359638eac7921420c155eb1946ead52cca31d7ecc2bf589671R112-R113) to support the new array handling of topic and offsets  

## Default partitioner
Another change is the update of the [DefaultPartitioner](https://kafka.js.org/docs/migration-guide-v2.0.0#producer-new-default-partitioner)
By default the JavaCompatiblePartitioner is used but if not explicitly set it will log a warning
We are setting it by default (if not already set) [here](https://github.com/Workable/orka/pull/384/files#diff-7a49104279a769359638eac7921420c155eb1946ead52cca31d7ecc2bf589671R37)
If for some reason someone needs to use the old partitioner (LegacyPartitioner), it can be set through the `.withKafkaOptions()`
e.g `.withKafka({ createPartitioner: Partitioners.LegacyPartitioner })`

manoskouvarakis commented on src/initializers/kafka/kafka.ts:
```diff
@@ -35,6 +34,7 @@ export default class OrkaKafka {
       authenticationTimeout
     });
 
+    if (options && !options.createPartitioner) options.createPartitioner = Partitioners.DefaultPartitioner;
```
Can we add the ability to define the partitioner in docs too? 
https://github.com/Workable/orka/blob/master/docs/integrations/kafka.md

manoskouvarakis reviewed (approved):
🚀

nikostoulas commented:
@kalyvasio Can we please also add the breaking changes here: https://workable.github.io/orka/#migrating-from-orka--3x-to-4x?

nikostoulas commented on src/initializers/kafka/kafka.ts:
```diff
@@ -179,7 +179,6 @@ function getAuthOptions(options: {
-
   const { username, password } = options.sasl || {};
-  if (username && password) return { sasl: options.sasl, ssl: options.ssl };
+  if (username && password) return { sasl: {...options.sasl }, ssl: options.ssl };
```
❓ Was this needed?

kalyvasio replied:
No it was a leftover from a reverted refactoring...Fixed in https://github.com/Workable/orka/pull/384/commits/e1bc8c05291ae760a6622fc09e8f96df986b5131

nikostoulas commented on src/utils.ts:
```diff
@@ -25,7 +25,7 @@ export const nodeVersionGreaterThanEqual = (requestedVersion: string, version =
 
 export function appendHeadersFromStore(
-  properties: { headers?: Record<string, string | Buffer | undefined> },
+  properties: { headers?: { [key: string]: Buffer | string | (Buffer | string)[] | undefined } },
```
❓ Was the previous complaining?

kalyvasio replied:
Yes but apparently for a different reason
Addressed here https://github.com/Workable/orka/pull/384/commits/c0576641c497705f62eabcac01d207b66903e27d

klesgidis commented on src/initializers/kafka/kafka.ts:
```diff
@@ -109,11 +109,11 @@ export default class OrkaKafka {
         .map(async ({ groupId, topic, oldGroupId }) => {
-          const offsets = await admin.fetchOffsets({ groupId, topic, resolveOffsets: false });
-          if (offsets.every(({ offset }) => offset === '-1')) {
+          const offsets = (await admin.fetchOffsets({ groupId, topics: [topic], resolveOffsets: false }))[0];
```
💅 
```suggestion
          const [offsets] = await admin.fetchOffsets({ groupId, topics: [topic], resolveOffsets: false });
```

klesgidis commented on src/initializers/kafka/kafka.ts:
```diff
@@ -109,11 +109,11 @@ export default class OrkaKafka {
             // groupId is not configured
-            const oldOffsets = await admin.fetchOffsets({ groupId: oldGroupId, topic, resolveOffsets: false });
-            const knownOffsets = oldOffsets?.filter(o => o.offset !== '-1');
+            const oldOffsets = (await admin.fetchOffsets({ groupId: oldGroupId, topics: [topic], resolveOffsets: false }))[0];
```
same

antonakospanos commented on src/initializers/kafka/index.ts:
```diff
@@ -39,7 +39,7 @@ function updateSend(kafka: Kafka, config: { traceHeaderName: string }) {
       appendHeadersFromStore(message, getRequestContext(), config);
       if (!message.key && message.headers[traceHeaderName]) {
-        message.key = message.headers[traceHeaderName];
+        message.key = message.headers[traceHeaderName] as Buffer | string | null;
```
Shouldn't accept `number` keys too?
```suggestion
        message.key = message.headers[traceHeaderName] as Buffer | string | number | null;
```

nikostoulas commented on src/typings/kafka.d.ts:
```diff
@@ -6,7 +6,15 @@ export interface KafkaConfig {
   };
   sasl?: {
-    mechanism: 'plain' | 'scram-sha-256' | 'scram-sha-512';
+    mechanism: 'plain';
+    username: string;
+    password: string;
+  } | {
+    mechanism: 'scram-sha-256';
+    username: string;
+    password: string;
+  } | {
+    mechanism: 'scram-sha-512';
```
❓ Is this equivalent with the previous? Or did sth change here?

kalyvasio replied:
It is equivalent...We only changed this because we had some issues with type inference in the Kafka Config [here](https://github.com/Workable/orka/blob/e1bc8c05291ae760a6622fc09e8f96df986b5131/src/initializers/kafka/kafka.ts#L88)

nikostoulas commented:
> @kalyvasio Can we please also add the breaking changes here: https://workable.github.io/orka/#migrating-from-orka--3x-to-4x?

FYI here https://github.com/Workable/orka/blob/915d882a915887b1cad7ed6a9d1a0e8260f9522e/docs/index.md?plain=1#L77 is the place to add the changes

kalyvasio commented:
> > @kalyvasio Can we please also add the breaking changes here: https://workable.github.io/orka/#migrating-from-orka--3x-to-4x?
> 
> FYI here
> 
> https://github.com/Workable/orka/blob/915d882a915887b1cad7ed6a9d1a0e8260f9522e/docs/index.md?plain=1#L77
> 
> is the place to add the changes

I have updated the docs and I have added the usage for the Default Partitioner in the Kafka md in [this commit](https://github.com/Workable/orka/pull/384/commits/3b9ec03b4fbac83aa5132daa88dcc0269c4fcc74)
Please take a look and tell me if you want something different

nikostoulas commented on docs/index.md:
```diff
@@ -75,3 +75,8 @@ ObjectId using new mongoose.
 - Methods kafka.send that was deprecated is removed. Use kafka.producer.send instead
 - Mongoose migrated to v7 which has breaking changes see [here](https://mongoosejs.com/docs/7.x/docs/migrating_to_7.html)
+
+### Migrating from orka  4.x to 5.x
```
@kalyvasio this is still migrating from 3.x to 4.x as we haven't gone out of beta yet. Use one line for the kafka changes

nikostoulas commented on docs/integrations/kafka.md:
```diff
@@ -242,3 +242,24 @@ You also need to specify the new group ids in your consumers while previously yo
 
 If you are using the renameGroupIds (before creating your consumers) your consumers will continue reading messages from the offset specified from the old groupId regardless if you set the fromBeginning configuration. FromBeginning configuration will be used if the groupId, topic is not found in kafka.
+
+## Migrating from orka < 5.x.x
```
```suggestion
## Migrating from orka < 4.x.x
```

nikostoulas commented on docs/integrations/kafka.md:
```diff
@@ -242,3 +242,24 @@ You also need to specify the new group ids in your consumers while previously yo
+The only possible implication is the update of the DefaultPartitioner (https://kafka.js.org/docs/migration-guide-v2.0.0#producer-new-default-partitioner). This may case
+issues in partition-aware environments, since some messages may be produced in different partitions than they would with the previous default partitioner.
+
+By default, Orka ^5.x.x will use the new DefaultPartitioner (previously named JavaCompatiblePartitioner).
```
```suggestion
By default, Orka ^4.x.x will use the new DefaultPartitioner (previously named JavaCompatiblePartitioner).
```
